language/vm/vm-runtime/vm-runtime-types/src/value.rs

```diff
@@ -161,6 +161,7 @@ impl ValueImpl {
                 Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR).with_message(msg))
             }
             ValueImpl::PromotedReference(reference) => reference.copy_value(),
+            ValueImpl::Struct(s) => s.copy_value(),
             _ => Ok(Value(self.clone())),
```

Comment:
Might be worthwhile to use exhaustive pattern matching here to ensure that every non-reference value is doing a deep copy instead of a shallow RC `clone`. E.g., I think the `Vector` bug is due to the fact that we didn't do this for `ValueImpl::NativeStruct` here.

Reply:
yes you are right!
It's funny that I thought about it and convinced myself we were good because I mentally run the match but forgot the vector is behind NativeStruct. I spent too much time lately to work with vector at the top level value and lost track of where it was.
Thanks so much as it kept bugging me for reasons I could not figure out and sure enough... :)

Reply:
Yeah, vector hiding behind native struct is sneaky... Hopefully, making it a first-class type (and perhaps eliminating `NativeStruct` altogether) as we are planning to do will help with this.